src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -533,6 +541,46 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
             location=node.location,
         )
 
+    def _extract_promoted_tuple_type(
+        self,
+        node_type_ls: list,
+        mask_type: ct.FieldType,
+        return_type: None | ct.FieldType | ct.ScalarType,
+    ) -> ct.FieldType:
+        for element in node_type_ls:
+            if isinstance(element, ct.TupleType):
+                return self._extract_promoted_tuple_type(element.types, mask_type, return_type)
+            else:
+                if return_type is None:
+                    return_type = element
+                else:
+                    return_type = type_info.promote(return_type, element)
+                    if isinstance(return_type, ct.ScalarType) or not all(
+                        item in return_type.dims for item in mask_type.dims
+                    ):
```

Comment:
If this condition could be built into `promote_to_mask_type` instead (or extracted into another function), then the whole `for` level `else` branch could be written

```python
            else:
                return_type = type_info.promote(return_type, element) if return_type else element
                return_type = self._promote_to_mask_type(mask_tye, return_type)  # or new function if preferred
```

Which significantly reduces the brain power required to follow the control flow (for me at least).